Fix #18272 TextInput.setNativeProps({text: ''}) to work

magicien opened the pull request:
## Motivation

Fix #18272. Calling textInputRef.setNativeProps({text: ''}) or textInputRef.clear() should clear the text input.  

## Test Plan

- All tests of `yarn run test` are passed
- Test with [the sample app](https://github.com/magicien/react-native-textinput-clear).
    - TextInput.clear() and TextInput.setNativeProps({ text: '***' }) worked
    - When clear() or setNativeProps() called, onChange/onChangeText wasn't called
        - Same behavior as react 0.53.0
    - When non-string values are given to `setNativeProps({text: ___})`, its behavior is the same as react 0.53.0.
        - Value Type | Result
          ---------- | ------------
          null       | same as empty string ''
          undefined  | nothing changes
          number     | throw error
          function   | throw error
          object     | throw error
    - When clear() or setNativeProps() called, attributed text keeps the attributes
    - When `value` prop is set, the text can't be changed

### react 0.54.0 (with the bug)

- `clear()` doesn't work from the second time
- `setNativeProps({text '***'})` doesn't work from the second time
- Even when `value` prop is set, you can change the text

![ScreenShot_0.54.0](https://raw.githubusercontent.com/magicien/react-native-textinput-clear/master/screenshot/0.54.0_test.gif)

### with bugfix

- `clear()` works every time
- `setNativeProps({text '****'})` works every time

![ScreenShot_Clear_1](https://raw.githubusercontent.com/magicien/react-native-textinput-clear/master/screenshot/clear_test_1.gif)

![ScreenShot_Clear_2](https://raw.githubusercontent.com/magicien/react-native-textinput-clear/master/screenshot/clear_test_2.gif)

### Attributed text with bugfix

- The text keeps the attributes (font family, size, color, text align)

![ScreenShot_Slider](https://raw.githubusercontent.com/magicien/react-native-textinput-clear/master/screenshot/attributed_text_test.gif)

### `value` prop with bugfix

- If `value` prop is set, the text should not be changed

![ScreenShot_Value](https://raw.githubusercontent.com/magicien/react-native-textinput-clear/master/screenshot/value_test.gif)

## Release Notes

[IOS] [BUGFIX] [TextInput] - Fix TextInput.clear() and TextInput.setNativeProps({text: ''}) to work

Leeds-eBooks commented:
What is the status of this? The regression fixed by this PR is holding us back from publishing our app so I'm keen to help out in any way I can!

magicien commented:
Thank you! I was testing my changes.
I'm going to update this PR and remove "WIP" soon!

magicien commented:
I added Test Plan and removed "WIP" from the title.
How can I solve the CI error? It seems to have nothing to do with this change.

magicien commented:
I added an attributed text test. (Thank you @reyalpsirc!)
And it seems that CI errors are fixed.

magicien commented:
@shergin Thank you for your review!
I looked into the code again, and I guess the problem is that the value of ShadowView is not sync with user input.
If input events are handled correctly, `text` property doesn't need to be added to UIView instance.
Let me think how to fix it.

magicien commented:
I removed `text` property from RCTBaseTextInputView and added `text` accessor to RCTBaseTextInputShadowView.
When `text` native property (or `value` prop of TextInput component) is explicitly set, `text` should be compared with the current local text (`_localAttributedText`) [to check if it needs to update](https://github.com/facebook/react-native/blob/15bc6d1e0b6f14ed42100515fef7a26419e71b0e/Libraries/Text/TextInput/RCTBaseTextInputShadowView.m#L141).

This change will also fix #18389 (TextInput value not binding on IOS).

edo1493 commented:
News?

magicien commented:
This pull request is ready for review.
@shergin Could you please check it?

albert-carreras commented:
Any news? When can someone merge this (maybe also cherry pick it for 0.54.5 or 0.55?)

StrictlySkyler commented:
Bump

luco commented:
Updates? this fixes my issued and should be merged.

albert-carreras commented:
This regression fix has been open for 20 days... It's not that big of a PR either, why is no one merging?

shergin commented on Libraries/Text/TextInput/RCTBaseTextInputShadowView.m:
```diff
@@ -101,6 +102,17 @@ - (void)invalidateContentSize
+- (void)setText:(NSString *)text
+{
+  _text = text;
+  _previousAttributedText = _localAttributedText;
```
I think we simply have to call `[self dirtyLayout];` here.

ptlu replied:
@shergin I've tried this out.  There is a still a bug where when iOS autocompletes it still doesn't clear.  Something happens with the autocomplete where it blocks the clear.

kevinresol replied:
That doesn't work.

kevinresol replied:
just to avoid any ambiguity:
`_previousAttributedText = _localAttributedText;` works

jason-feng replied:
@shergin @kevinresol any updates?

stueynet replied:
@shergin you requested changes on this 10 weeks ago. Any chance you could approve it as is? This issue causes Android text boxes to be unusable. Its a major blocker for many production apps.

hramos replied:
This PR is waiting for @magicien to follow up to @shergin's feedback, or alternatively for @shergin to sign off on the PR to be merged as-is.

It's not clear to me that this can be merged just yet, as it seems like @shergin asked for `[self dirtyLayout]` to be called after `_previousAttributedText = _localAttributedText;` and the feedback has not been addressed.

kevinresol replied:
What I understand from @shergin 's comment is to replace `_previousAttributedText = _localAttributedText;` with `[self dirtyLayout];` and I have already stated clearly that it does not work.

magicien replied:
@shergin @kevinresol @hramos Apologies for taking so long to reply.
I would like to make sure what @shergin means.
@shergin, do you mean this code?

```
- (void)setText:(NSString *)text
{
  _text = text;
  _previousAttributedText = _localAttributedText;
  [self dirtyLayout];
}
```

I'm wondering if it needs to call `dirtyLayout` because `dirtyLayout` is called after `setText` is called. Calling `dirtyLayout` could be safer, though.
If the above code is the expected one, I will update this PR.

albert-carreras replied:
Would be nice to have a reply/merge @shergin to resolve this. IMO, the PR is working as it is, and the suggested change doesn't seem to work.

kevinresol replied:
@shergin : a friendly reminder to review this, thanks.

shergin replied:
@magicien If you can confirm that `_previousAttributedText = nil;` works and update the diff, I would land this.

Noitidart commented:
The same is happening for Android on very few devices.

* Simple test case with snack expo - https://github.com/HandlebarLabs/code-and-coffee/issues/15#issuecomment-379606414

* A more advanced test case with snack expo - https://github.com/HandlebarLabs/code-and-coffee/issues/15#issue-302470239

maraujop commented:
this bug has been around since 0.54, is this fix going to be merged in 0.56 (April) ? 

I find it hard to believe a bug like this makes it to React Native and stays longer than 2 versions around, with a PR submitted almost 2 months ago. For us this bug is a total stopper, thanks.

Noitidart commented:
Created an issue for Android - https://github.com/facebook/react-native/issues/19085 - it's similar but not the same.

pualxiao commented:
https://github.com/pualxiao/React-Native-TextInput-Bug-Fix

Noitidart commented:
Thank you @pualxiao for sharing that copy paste fix!! Was very much needed until this PR gets merged.

hramos commented:
@maraujop I know it can be frustrating, but if you look at the history of this PR, you can see it was reviewed by @shergin several times. He asked for changes back in April 11 but @magicien has not responded yet.

Noitidart commented:
It would be nice if it can be group effort. Lots of times the changes requested are harder for the submitter. Ideally the person requesting the changes can help out, they know what they want done.  @magicien already did an extremely amazing job, finding there was a bug, creating a videos to show others before and after, and a possible fix.

hramos commented:
@Noitidart I agree with the sentiment, and hopefully it's something we can do more often once we have fewer PRs waiting for review.

assafb81 commented:
@shergin,
any updates?

minhchienwikipedia commented:
I had create new Text Input for fix this bug right now, everybody can use it before RN fix it.
https://github.com/agiletechvn/react-native-text-input-enhance

anhtuank7c commented:
@minhchienwikipedia You saved my day. Thank you so much.

sasayahiromu commented:
Is this pull request merged in 0.56??
Or, Is there any workaround??

kevinresol commented:
> @Noitidart I agree with the sentiment, and hopefully it's something we can do more often once we have fewer PRs waiting for review.

Understand that, but I suppose this bug should deserve a very high priority.

hramos commented:
@sasayahiromu this PR is open and has not been merged to master, so it's not in any release.

hramos commented:
@kevinresol feel free to open your own PR. It does not look like the author of this PR, @magicien, is actively working on it, as they have not responded to @shergin's feedback.

kevinresol commented:
> as they have not responded to @shergin's feedback.

But I did

gianpaj commented:
Can anyone pick up the work, please? Validation on input texts are basically useless on iOS.

Thanks

bbeckk commented:
Simply unfocusing the textInput solves the problem

```
<View style={styles.textInputStyle}>

    <TextInput onChangeText={(text) => this.setState({searchText: text}, () => {
                   this.state.searchText.length > 1 ? this._getSearchedData() : null
               })}
               value={this.state.searchText}
               placeholder='Search restaurant or cuisine'
    />

    <TouchableOpacity  onPress={() => {
                    this.setState({searchText: ''}),
                    Keyboard.dismiss()
                }}>
    <Icon name="md-close-circle" style={{color: 'gray', fontSize: 15,}}/>
</TouchableOpacity>

</View>
```

kevboh commented:
Hello! If the original author of a PR does not return to answer feedback, is there anything the community can do to help get that PR merged? It's sort of silly to leave huge regressions like this alive in the wild just because a contributor ghosted on five lines of code.

kientux commented:
This PR also fixes my issue #20224.
I'm new to React Native and already surprised by the number of very simple bugs 😬

magicien commented:
@shergin Could you please check my comment on your review?

hramos commented:
@kevboh you can always submit your own PR, making sure to incorporate the feedback brought up here.

kevinresol commented:
> @kevboh you can always submit your own PR, making sure to incorporate the feedback brought up here.

The original author has already returned and commented.  @hramos @shengmin Please review

peacechen commented:
This PR has languished for far too long.  Binding the TextInput value to state is critical for form validation.  Please prioritize merging this into a stable release.

hramos commented:
I've already left my thoughts on the status of this PR. No need to @-mention me any more.

stueynet commented:
@hramos

kevinresol commented:
@shergin please follow up

kevinresol commented:
@shergin please follow up

magicien commented:
I updated the diff and added [the comment](https://github.com/facebook/react-native/pull/18278#discussion_r211106330).
It seems that there is something wrong with the test environment.

kevinresol commented:
any updates?

ngandhy commented:
@hramos - See: https://github.com/facebook/react-native/pull/18278#discussion_r210798429 . The requested changes by @shergin 15 days ago was made by original author 12 days ago. Can we get this regression merged now? If @shergin is still MIA on the other project this may sit around indefinitely

Thanks!

kevinresol commented:
@shergin please follow up, thank you very much.

hramos commented:
@rowinbot it's imported. Once the reviewer accepts the changes, it will land internally and eventually sync out to github.

hramos commented:
I've turned off these types of notifications from the bot for now.

We check the thread for an existing comment from the bot before adding a new comment. I'll need to look into why this failed on this commit.
